File: apps/web/types/procurement.ts
export type ProcurementKind = 'proposal' | 'plan' | 'acceptance'
export type ProcurementStatus = 'draft' | 'submitted' | 'approved' | 'rejected' | 'completed' | 'archived'

export interface ProcurementItem {
  id: string
  document_id: string
  document_title?: string | null
  document_number?: string | null
  document_status?: string | null
  procurement_kind: ProcurementKind
  reference_number?: string | null
  title_summary?: string | null
  requesting_unit?: string | null
  estimated_value?: string | number | null
  currency: string
  requested_date?: string | null
  status: ProcurementStatus
  notes?: string | null
  created_at: string
  updated_at: string
}

export interface ProcurementInput {
  document_id: string
  procurement_kind: ProcurementKind
  reference_number?: string
  title_summary?: string
  requesting_unit?: string
  estimated_value?: string | number | null
  currency?: string
  requested_date?: string
  status: ProcurementStatus
  notes?: string
}

export interface ProcurementListFilters {
  q?: string
  document_id?: string
  procurement_kind?: ProcurementKind | ''
  reference_number?: string
  requesting_unit?: string
  status?: ProcurementStatus | ''
  requested_date_from?: string
  requested_date_to?: string
  sort_by?:
    | 'created_at'
    | 'updated_at'
    | 'reference_number'
    | 'procurement_kind'
    | 'requesting_unit'
    | 'status'
    | 'requested_date'
    | 'estimated_value'
  sort_dir?: 'asc' | 'desc'
  limit?: number
  offset?: number
}

export interface ProcurementListResponse {
  items: ProcurementItem[]
  total: number
  limit: number
  offset: number
}

File: apps/api/app/schemas/procurement.py
from datetime import date, datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, Field


ProcurementKind = Literal["proposal", "plan", "acceptance"]
ProcurementStatus = Literal["draft", "submitted", "approved", "rejected", "completed", "archived"]


class ProcurementBase(BaseModel):
    document_id: str
    procurement_kind: ProcurementKind
    reference_number: str | None = Field(default=None, max_length=128)
    title_summary: str | None = None
    requesting_unit: str | None = Field(default=None, max_length=255)
    estimated_value: Decimal | None = Field(default=None, ge=0)
    currency: str = Field(default="VND", max_length=8)
    requested_date: date | None = None
    status: ProcurementStatus = "draft"
    notes: str | None = None


class ProcurementCreateRequest(ProcurementBase):
    pass


class ProcurementUpdateRequest(BaseModel):
    procurement_kind: ProcurementKind | None = None
    reference_number: str | None = Field(default=None, max_length=128)
    title_summary: str | None = None
    requesting_unit: str | None = Field(default=None, max_length=255)
    estimated_value: Decimal | None = Field(default=None, ge=0)
    currency: str | None = Field(default=None, max_length=8)
    requested_date: date | None = None
    status: ProcurementStatus | None = None
    notes: str | None = None


class ProcurementRead(BaseModel):
    id: str
    document_id: str
    document_title: str | None = None
    document_number: str | None = None
    document_status: str | None = None
    procurement_kind: str
    reference_number: str | None = None
    title_summary: str | None = None
    requesting_unit: str | None = None
    estimated_value: Decimal | None = None
    currency: str
    requested_date: date | None = None
    status: str
    notes: str | None = None
    created_at: datetime
    updated_at: datetime


class ProcurementListResponse(BaseModel):
    items: list[ProcurementRead]
    total: int
    limit: int
    offset: int
